# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFrameMessageManager.cpp
# Commit: 2bc5949902ac
# Full Hash: 2bc5949902acc042205fc7e85260bbbc37dbedc7
# Author: Matthew Gaudet <mgaudet@mozilla.com>
# Date: 2021-09-18 09:46:57
# Regressor Bug: 1688788
# File Overlap Count: 1
# Description:
#   Bug 1688788 - Convert ScriptPreloader to manage Stencils, not Scripts. r=kmag,tcampbell
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D113013
# ==============================================================================

diff -r e660e2c3c3d1 -r 2bc5949902ac dom/base/nsFrameMessageManager.cpp
--- a/dom/base/nsFrameMessageManager.cpp	Sat Sep 18 01:29:45 2021 +0000
+++ b/dom/base/nsFrameMessageManager.cpp	Sat Sep 18 01:29:45 2021 +0000
@@ -19,6 +19,7 @@
 #include "chrome/common/ipc_channel.h"
 #include "js/CallAndConstruct.h"  // JS::IsCallable, JS_CallFunctionValue
 #include "js/CompilationAndEvaluation.h"
+#include "js/experimental/JSStencil.h"
 #include "js/JSON.h"
 #include "js/PropertyAndElement.h"  // JS_GetProperty
 #include "js/SourceText.h"
@@ -1254,17 +1255,17 @@
   JSContext* cx = jsapi.cx();
 
   JS::CompileOptions options(cx);
-  ScriptPreloader::FillCompileOptionsForCachedScript(options);
+  ScriptPreloader::FillCompileOptionsForCachedStencil(options);
   options.setFileAndLine(url.get(), 1);
   options.setNonSyntacticScope(true);
 
-  JS::Rooted<JSScript*> script(cx);
+  RefPtr<JS::Stencil> stencil;
   if (useScriptPreloader) {
-    script =
-        ScriptPreloader::GetChildSingleton().GetCachedScript(cx, options, url);
+    stencil =
+        ScriptPreloader::GetChildSingleton().GetCachedStencil(cx, options, url);
   }
 
-  if (!script) {
+  if (!stencil) {
     nsCOMPtr<nsIChannel> channel;
     NS_NewChannel(getter_AddRefs(channel), uri,
                   nsContentUtils::GetSystemPrincipal(),
@@ -1311,18 +1312,25 @@
       return;
     }
 
-    script = JS::Compile(cx, options, srcBuf);
-    if (!script) {
+    stencil = JS::CompileGlobalScriptToStencil(cx, options, srcBuf);
+    if (!stencil) {
       return;
     }
   }
 
-  MOZ_ASSERT(script);
+  MOZ_ASSERT(stencil);
+
+  JS::Rooted<JSScript*> script(
+      cx, JS::InstantiateGlobalStencil(cx, options, stencil));
+  if (!script) {
+    return;
+  }
+
   aScriptp.set(script);
 
   if (useScriptPreloader) {
-    ScriptPreloader::GetChildSingleton().NoteScript(url, url, script,
-                                                    isRunOnce);
+    ScriptPreloader::GetChildSingleton().NoteStencil(url, url, stencil,
+                                                     isRunOnce);
 
     // If this script will only run once per process, only cache it in the
     // preloader cache, not the session cache.